mlir/lib/Transforms/Utils/LoopInvariantCodeMotionUtils.cpp

```diff
@@ -58,62 +60,256 @@ static bool canBeHoisted(Operation *op,
+    for (Operation &opInner : region.getOps())
+      gatherResourceConflicts(loopLike, &opInner, resourceConflicts);
+}
+
+size_t mlir::moveLoopInvariantCode(
+    LoopLikeOpInterface loopLike,
+    function_ref<bool(Value, Region *)> isDefinedOutsideRegion,
+    function_ref<bool(Operation *, Region *)> shouldMoveOutOfRegion,
+    function_ref<void(Operation *, Region *)> moveOutOfRegion) {
+  size_t numMovedTotal = 0;
+
+  // TODO: see if this can be spec'd in a meaningful way to add back later.
+  //
+  // check that the loop isn't dead
+  auto isDead = loopLike.isZeroTrip();
+
+  LDBG() << "Loop " << OpWithFlags(loopLike.getOperation(), OpPrintingFlags().skipRegions())
+    << " has constant bounds and steps? isZeroTrip()? " << (isDead.has_value() ? (isDead.value() ? "YES, YES" : "YES, NO") : "NO, NULL");
+
+  if (isDead.has_value() && isDead.value())
+    return numMovedTotal;
+
+  int numMoved = 0;
+
+  do {
+    // reset value for iteration
+    numMoved = 0;
+
+    // Go through loop body and map out resource usages.
+    // op->regions are essentially merged sequentially.
+    // E.g., an if's "then" and "else" regions are treated like one
+    // continuous region --> need to add fork checking.
+    //
+    // loop "do" and "then" regions also merged.
+    DenseMap<TypeID, std::pair<bool, MemoryEffects::EffectInstance>>
+        resourceConflicts;
+    gatherResourceConflicts(loopLike, loopLike.getOperation(), resourceConflicts);
+
+    auto regions = loopLike.getLoopRegions();
+    for (Region *region : regions) {
+      LDBG() << "Original loop:\n" << *region->getParentOp();
+
+      std::queue<Operation *> worklist;
+      // Add top-level operations in the loop body to the worklist.
+      for (Operation &op : region->getOps())
+        worklist.push(&op);
+
+      auto definedOutside = [&](Value value) {
+        return isDefinedOutsideRegion(value, region);
+      };
+
+      while (!worklist.empty()) {
+        Operation *op = worklist.front();
+        worklist.pop();
+        // Skip ops that have already been moved. Check if the op can be hoisted.
+        if (op->getParentRegion() != region)
+          continue;
+
+        LDBG() << "Checking op: "
+              << OpWithFlags(op, OpPrintingFlags().skipRegions());
+
+        bool noMemoryConflicts =
+            isMemoryEffectFree(op) ||
+            !mayHaveMemoryEffectConflict(op, resourceConflicts);
+
+        if (!noMemoryConflicts || !shouldMoveOutOfRegion(op, region) ||
+            !canBeHoisted(op, definedOutside))
+          continue;
+
+        LDBG() << "Moving loop-invariant op: " << *op;
+        moveOutOfRegion(op, region);
+        ++numMoved;
+
+        // Since the op has been moved, we need to check its users within the
+        // top-level of the loop body.
+        for (Operation *user : op->getUsers())
+          if (user->getParentRegion() == region)
+            worklist.push(user);
+      }
+    }
+
+    numMovedTotal += numMoved;
```

Comment:
```suggestion
    numMovedTotal += numMoved;
    LDBG() << "Finishing LICM iteration " << iteration++ << " moved " << numMoved << " ops, total is now " << numMovedTotal;
```
(need to add `int iteration = 0;` before the loop as well)